.azure-pipelines.yml

```diff
@@ -0,0 +1,248 @@
+name: $(Build.SourceBranch)-$(date:yyyyMMdd)$(rev:.r)
+trigger:
+  branches:
+    include:
+    - 'master'
+  tags:
+    include:
+    - '*'
+
+jobs:
+- job: Build
+  strategy:
+    matrix:
+      windows-stable:
+        imageName: 'vs2017-win2016'
+        rustup_toolchain: stable
+      windows-stable-release:
+        imageName: 'vs2017-win2016'
+        rustup_toolchain: stable
+        build_type: release
+      linux-stable-release:
+        imageName: 'ubuntu-16.04'
+        rustup_toolchain: stable
+        build_type: release
+      mac-stable-release:
+        imageName: 'macos-10.14'
+        rustup_toolchain: stable
+        build_type: release
+      linux-stable:
+        imageName: 'ubuntu-16.04'
+        rustup_toolchain: stable
+      mac-stable:
+        imageName: 'macos-10.14'
+        rustup_toolchain: stable
+      mac-beta:
+        imageName: 'macos-10.14'
+        rustup_toolchain: beta
+      mac-nightly:
+        imageName: 'macos-10.14'
+        rustup_toolchain: nightly
```

Comment:
I personally like a stylistic strategy like https://github.com/rust-lang/git2-rs/blob/1d33b290e034d0d85ac0e4bc0ac64cebfa8cfbc7/azure-pipelines.yml where one-off jobs have their own `job` entry and groups of jobs have their own `matrix` entry. I find that putting everything into the same matrix makes for some pretty gnarly `condition` entires and lots of if-this-else-that in the config.

This is totally fine to be clear, just figured I'd mention that this may be a way to simplify the config below.